pkg/service/repair/service.go

```diff
@@ -454,71 +462,73 @@ func (s *Service) SetParallel(ctx context.Context, clusterID uuid.UUID, parallel
 	return errors.Wrap(err, "update db")
 }
 
-type sizeSetter interface {
-	SetSize(size int)
-}
-
-type intensityHandler struct {
-	taskID           uuid.UUID
-	runID            uuid.UUID
-	logger           log.Logger
-	maxHostIntensity map[string]Intensity
-	intensity        *atomic.Int64
-	maxParallel      int
-	parallel         *atomic.Int64
-	poolController   sizeSetter
-}
-
-const (
-	maxIntensity     Intensity = 0
-	defaultIntensity Intensity = 1
-	defaultParallel            = 0
-	chanSize                   = 10000
-)
-
-// SetIntensity sets the value of '--intensity' flag.
-func (i *intensityHandler) SetIntensity(ctx context.Context, intensity Intensity) {
-	i.logger.Info(ctx, "Setting repair intensity", "value", intensity, "previous", i.intensity.Load())
-	i.intensity.Store(int64(intensity))
-}
-
-// SetParallel sets the value of '--parallel' flag.
-func (i *intensityHandler) SetParallel(ctx context.Context, parallel int) {
-	i.logger.Info(ctx, "Setting repair parallel", "value", parallel, "previous", i.parallel.Load())
-	i.parallel.Store(int64(parallel))
-	if parallel == defaultParallel {
-		i.poolController.SetSize(i.maxParallel)
-	} else {
-		i.poolController.SetSize(parallel)
-	}
-}
-
-func (i *intensityHandler) ReplicaSetMaxIntensity(replicaSet []string) Intensity {
-	out := NewIntensity(math.MaxInt)
-	for _, rep := range replicaSet {
-		if ranges := i.maxHostIntensity[rep]; ranges < out {
-			out = ranges
+func shouldBatchRanges(session gocqlx.Session, clusterID, taskID, runID uuid.UUID) (bool, error) {
+	prevIDs, err := getAllPrevRunIDs(session, clusterID, taskID, runID)
+	if err != nil {
+		return false, err
+	}
+	if len(prevIDs) == 0 {
+		return true, nil
+	}
+
+	q := qb.Select(table.SchedulerTaskRun.Name()).Columns(
+		"status",
+	).Where(
+		qb.Eq("cluster_id"),
+		qb.Eq("type"),
+		qb.Eq("task_id"),
+		qb.Eq("id"),
+	).Query(session)
+	defer q.Release()
+
+	var status string
+	for _, id := range prevIDs {
+		err := q.BindMap(qb.M{
+			"cluster_id": clusterID,
+			"type":       "repair",
+			"task_id":    taskID,
+			"id":         id,
+		}).Scan(&status)
+		if err != nil {
+			return false, errors.Wrap(err, "get prev run status")
+		}
+		// Fall back to no-batching when some of the previous runs:
+		// - finished with error
+		// - got out of scheduler window
```

Comment:
Can you elaborate on the second statement ? 
What do you mean by "got out of scheduler window" ? Is it connected to maintenance window ? If so, then why is it the indicator to not batch ?

Reply:
As described in https://github.com/scylladb/scylla-manager/issues/3792#issuecomment-2112151831, I was worried that batching combined with a maintenance window set to many, small time windows (e.g. repair can run for between `X:45-X:59` where X can be any hour of any day), could result in repair not making any progress as repair would be interrupted by going out of maintenance window before finishing any repair job. This could be especially true for tables with only one replica set (e.g. fully replicated table), where batching could send 10% of all ranges (so usually` #node_cnt * 256 / 10` token ranges) in a single repair job.

Having said that, after playing with SCT performance tests, I plan to change batching from `max(intensity, 10% of replica set ranges)` to `min(100% of replica set ranges, 10 * intensity)`, so the maintenance window argument will be weaker.

@karol-kokoszka do you think that we shouldn't worry about maintenance window and SM shouldn't fallback to non-batching approach when repair runs are interrupted by it?

Reply:
On the other hand, I think I will change batching to `min(100% of replica set ranges, 1000)`.

Sending all ranges owned by replica set is what asias suggested initially and this number might be big only for fully replicated tables, which won't happen in bigger (more than 9 nodes) clusters. Hard limit of 1000 makes more sense than 10 * intensity when running with low intensity setting (batching helps with non max intensity as well).
In real life, this limit won't really be hit anyway.

Reply:
Ok, let's leave the fallback on maintenance window.